Allow independent log processing pipelines (#4062)
open-telemetry/opentelemetry-specification: specification/logs/sdk.md

* [`SpanId`](./data-model.md#field-spanid)
* [`TraceFlags`](./data-model.md#field-traceflags)

> The SDK MAY offer a way to make a copy of `ReadWriteLogRecord`
> so that the user can set up independent log processing pipelines.

Reviewer: ```suggestion
The SDK MAY offer a way to make a copy of `ReadOnlyLogRecord`
so that the user can set up independent log processing pipelines.
```

I'm concerned that the use of "Write" in this sentence will be misleading. If the pipelines are independent, the record cannot be safely mutated.

I think the record should be finalized before it is given to the pipeline and it should be declared "immutable", so when we talk about independent pipelines, we should be talking about immutable data. However, it is possible for a processor to _change_ the data in a pipeline by constructing new immutable data.

Therefore, it seems we need two hooks at the end of a span or log record. There is a read-write hook, they are applied in order and all processors will see the outcome of read-write mutations. After that happens, the record becomes immutable and is passed to a second hook.

Considering the proposal in #4024, the two hooks are named:

- OnEnding() -- read/write, processors are called in order, all processors see changes resulting from these hooks
- OnEnd() -- read-only, processors are called in implementation-defined order; here the processor has an opportunity to change the data by constructing new data.

The "bug" in the specification is that Logs SDKs already have OnEnd() taking mutable data, having been combined with the functionality called OnStart() in the span SDK. We can fix the bug, I think, by deprecating the old, inconsistent names while leaving them for compatibility. Ignoring the word "End" purposefully, I would propose a consistent set of names like:

- OnStart() -- read/write hook for spans
- OnFinish() -- read/write hook for modifying finished spans / logs
- OnExport() -- read-only hook for exporting immutable spans / logs

where OnExport has a way to clone/copy so as to allow changes to the immutable data.

Author: The `ReadOnlyLogRecord` is a term defined in this document. It is only used by the processor. There is also `ReadOnlyLogRecord` which is used by exporters that do not need to be able to modify the log record.

> If the pipelines are independent, the record cannot be safely mutated.

The aim is so that the pipelines can be independent. It does not make sense from performance standpoint to make a deep copy of whole log record before passing to subsequent log record processors. Moreover most languages designed a log processor to accept a reference to a log record. Therefore for these languages it is not even possible to make a backwards-compatible change that would make pipelines independent by default.

>  I think the record should be finalized before it is given to the pipeline and it should be declared "immutable", so when we talk about independent pipelines, we should be talking about immutable data. However, it is possible for a processor to change the data in a pipeline by constructing new immutable data.

This is not how most of the existing stable Logs SDKs are designed. I aimed for a proposal which would be backwards compatible.

> Therefore, it seems we need two hooks at the end of a span or log record. There is a read-write hook, they are applied in order and all processors will see the outcome of read-write mutations. After that happens, the record becomes immutable and is passed to a second hook.

Why there is a need of a second hook? Why cannot user just use the first hook an not modify anything?

Author: > If we accept the statements in https://github.com/open-telemetry/opentelemetry-specification/pull/4030 and agree with your argument here, then the solution for the Span processor should look more like the solution here and not add a new hook (i.e., not add OnEnding() to overcome the "independence" problem there).

Can you elaborate what the solution do you mean. Take notice: 
- https://github.com/open-telemetry/opentelemetry-specification/issues/1089#issuecomment-2045556103
- https://github.com/open-telemetry/opentelemetry-specification/issues/1089#issuecomment-2061289086

decorate built-in processors for advanced scenarios such as enriching with
attributes.

The SDK SHOULD allow users to set up independent processing pipelines.
For instance, the user should be able to make a log record modification in
a processor that is not visible to and does not affect the behavior of other

> log record processors.

Author: What does adding a read-only hooks gives if the user can still use a hook which allows mutation?

> log record processors.

Reviewer: I understand how you propose to solve the problem, but this sounds to me like a defensive form of safety--for processors to voluntarily clone records--when the SDK could offer more automatic protection and potentially eliminate copies. I propose we examine a solution where the SDK provides a fanout processor that hides this detail and avoids cloning records when there is a single consumer.

> log record processors.

Author: > I propose we examine a solution where the SDK provides a fanout processor that hides this detail and avoids cloning records when there is a single consumer.

That is a good idea. An SDK could provide such Processor as it could be used across all languages. I will do my best to update the PR to reflect your proposal.

> log record processors.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4062/commits/ddeddbe6788b645968de73608f23039e201cfd56

> log record processors.

Author: After more thoughts I decided to propose an "isolated processor" instead https://github.com/open-telemetry/opentelemetry-specification/pull/4062/commits/69ef0f764963506c0dfb06601caf3ad73d67166e

the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
asynchronous processing) only reads are permitted.

It is implementation specific whether `logRecord` modification are
also applied to the subsequent registered processors.

> For instance, `logRecord` may be passed by value or by reference.

Author: See https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2126709559

> For instance, `logRecord` may be passed by value or by reference.

Author: > SDKs SHOULD provide an idiomatic FanoutProcessor for Logs/Tracing SDKs that calls children synchronously and in-order while ensuring they do not share mutable data.

I agree with this 

> Logs SDK OnEmit() mutations will be visible to the next processor to be consistent with Trace OnStart()

I will check if it is possible to keep the feature for sake of consistency even though it's pitfalls I outlined [here](https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2126709559). My main goal is to keep a possibility to have zero allocations when the log records are not modified (as this the the usual use case).

> For instance, `logRecord` may be passed by value or by reference.

Author: > SDKs SHOULD provide an idiomatic FanoutProcessor for Logs/Tracing SDKs that calls children synchronously and in-order while ensuring they do not share mutable data.

I agree with this

> For instance, `logRecord` may be passed by value or by reference.

Author: > Logs SDK OnEmit() mutations will be visible to the next processor to be consistent with Trace OnStart()

I will check if it is possible to keep the feature for sake of consistency even though it's pitfalls I outlined [here](https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2126709559). My main goal is to keep a possibility to have zero allocations when the log records are not modified (as this the the usual use case).

What are we going to do with the fact that the stable C++ Logs SDK already has taken a different route?

If we succeed to find a satisfying design for Go (and Rust), we could update the specification to make this requirement explicit. However, is it OK that C++ will behave differently?

For now, I will remove this change from the PR.

CC @open-telemetry/rust-maintainers @open-telemetry/cpp-maintainers @open-telemetry/technical-committee

> For instance, `logRecord` may be passed by value or by reference.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4062/commits/f3ced5cf401e1bc08153e8cd49a1320955a0f53f

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

#### Isolated processor

> This is a `LogRecordProcessor` decorator ensuring the log record passed to

Reviewer: This sounds good, though to me the term "decorator" sounds more applicable in some languages than others. I found the term is used twice in the OTel specification, once to describe the `ParentBased` sampler (where I see it as a mis-use of the term because `ParentBased` wraps up to 5 child samplers), and once to describe the LogsRecordProcessor, where it gives motivation for the PR here.

> The SDK MUST allow users to implement and configure custom processors and decorate built-in processors for advanced scenarios such as enriching with attributes.

Still, if anyone pushed back on the term "decorator", I would prefer to see us use terms like "language-specific mechanism to support safe and efficient composition of processor pipelines".

Author: Makes sense to me. PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4062/commits/7ee16e92915df6f1ce90b548dff2cab90d759ae6

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

#### Fan-out processor

This is a `LogRecordProcessor` composite ensuring the log record passed to
`OnEmit` of the configured `processors` do not share mutable data.
For example, the `OnEmit` implementation of the fan-out processor can make
a deep copy of the log record before passing it to each warpped processor.

**Configurable parameters:**

> * `processors` - independent log processors.

Author: I think that providing an "isolated processor" may be simpler and good-enough.
@jmacd, thoughts?

```suggestion
#### Isolated processor

This is a `LogRecordProcessor` decorator ensuring the log record passed to
`OnEmit` of the configured `processor` does not share mutable data with
subsequent registered processors.
For example, the `OnEmit` implementation of the isolated processor can make
a deep copy of the log record before passing it to the wrapped processor.

**Configurable parameters:**

* `processor` - log processor to be isolated.
```

> #### Fan-out processor
> 
> This is a `LogRecordProcessor` composite ensuring the log record passed to
> `OnEmit` of the configured `processors` do not share mutable data.
> For example, the `OnEmit` implementation of the fan-out processor can make
> a deep copy of the log record before passing it to each wrapped processor.
> 
> **Configurable parameters:**
> 
> * `processors` - independent log processors.

Author: I think that providing an "isolated processor" may be simpler and good-enough.
@jmacd, thoughts?

```suggestion
#### Isolated processor

This is a `LogRecordProcessor` decorator ensuring the log record passed to
`OnEmit` of the configured `processor` does not share mutable data with
subsequent registered processors.
For example, the `OnEmit` implementation of the isolated processor can make
a deep copy of the log record before passing it to the wrapped processor.

**Configurable parameters:**

* `processor` - log processor to be isolated.
```

Author: I decided to go with my proposal. I made this a separate commit to make reverting easier if needed.

https://github.com/open-telemetry/opentelemetry-specification/pull/4062/commits/69ef0f764963506c0dfb06601caf3ad73d67166e

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

> #### Isolated processor

Reviewer 2: Should this be immediately part of this stable document or start as experimental and go through an incubation phase?

> #### Isolated processor

Reviewer 3: Isolated processor name implies that some processing would be done by it, but in isolated mode. 
But based on the description, the only thing it should be doing is making log record immutable.

If it's the case, could it be made more obvious in the name? E.g. `Cloning Processor`, `Freezing Processor`, `Immuting Processor`?

Author: > Isolated processor name implies that some processing would be done by it, but in isolated mode.

The configured processor would be isolated mode.

I will rename it to `Isolating processor`. I think that "cloning" is implementation specific. Freezing and immuting are not correct names as the configured processor can stil modify the log record. It is just independent from other registered processors.

This is an implementation of `LogRecordProcessor` ensuring the log record

> passed to `OnEmit` of the configured `processor` does not share mutable data

Reviewer 3: ```suggestion
passed to `OnEmit` of this `processor` does not share mutable data
```

?

Author: I think the current way is more precise. Notice that the log processor passed to isolating processor may share mutable data and then it e.g. makes a clone and passes is to the configured processor to make sure that it won't be affected by any external mutations.

Reviewer 3: thanks, it makes sense now, along with configurable parameters section.

with subsequent registered processors.

> For example, the `OnEmit` implementation of the isolated processor can be
> a decorator that makes a deep copy of the log record before passing it to

Reviewer 3: nit:

maybe

```suggestion
For example, the `OnEmit` implementation of the isolated processor can 
make a deep copy of the log record before passing it to
```

Author: I just wanted to describe a possible implementation and I think that calling out a "decorator" here can be more descriptive. 

@jmacd, thoughts?

Author: Leaving as it is as I think calling out a decorator as an example is not bad nor misleading.
